ZOOKEEPER-3482: add unit tests for client SASL authentication over SSL

symat opened the pull request:
We had some issues about configuring Kerberos authentication with SSL, so I created some unit tests to verify that these features work together. The problem was originally reported on 3.5.5. There was some conflicts to backport the tests to 3.5, I will prepare a separate PR on that branch.

eolivelli commented:
please cherry pick to branch-3.6

anmolnar commented:
Thanks @symat . I need to take a closer look, because you added a passing unit test while I reported and issue from a running cluster where I was unable to use SASL+SSL in conjunction.

Either I did something wrong or your unit test does not cover the scenario properly, I have to figure it out first.

symat commented:
I agree, we should test this on a real server.
Also FYI: the same tests fail on 3.5.5 and I think you originally tested it there. I am still investigating, but I think you need to use slightly different configuration on 3.5.5 to make the SASL and SSL to work together.

symat commented:
FYI: I managed to make these tests to work also on 3.5.5. I uploaded it here https://github.com/symat/zookeeper/commit/6496325c1643a58a30e23d4f6a54734bc4a1245d (I think this is not something we would like to push to any apache branch).

Some system properties behave differently because of ZOOKEEPER-1467, ZOOKEEPER-3388 and ZOOKEEPER-3590 which tickets are on branch-3.5 and on master, but not on branches 3.5.5 nor on 3.5.6.

Now I am working on to try the different ZooKeeper versions with real kerberos and SSL.

symat commented:
The tests were successful using ZooKeeper 3.5.5 over SSL on real Hadoop cluster with MIT Kerberos.

I changed the code to reload the JAAS and kerberos configs properly between the tests. I also updated the documentation in git (plus updated the https://cwiki.apache.org/confluence/display/ZOOKEEPER/Client-Server+mutual+authentication wiki page). I think I am ready with this PR, @anmolnar @eolivelli PTAL.

(I also did the same changes on the other PR created for the 3.5 branch: https://github.com/apache/zookeeper/pull/1205)

anmolnar commented:
Great, committed to master and 3.6 Thanks @symat !
